core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java

```diff
@@ -154,6 +154,15 @@ public int offset() {
     return offset;
   }
 
+  /**
+   * Reset the backing array for this byte sequence object. This is useful for object re-use.
+   */
+  public void reset(byte[] data) {
```

Comment:
> I found MutableByteSequence.setArray after making these changes. Is ArrayByteSequence supposed to be immutable? There's no mention of it being so in the class and changes to the backing array are reflected in the object, so I'm thinking not.

I agree its not immutable since the backing array can easily be modified.   Adding this method seems useful, could add it to ArrayByteSequence and delete   MutableByteSequence.  Not sure how big of a change the would be, did not look at what uses MutableByteSequence

Reply:
> did not look at what uses MutableByteSequence

It's only used in RelativeKey and LocalityGroupUtil. I'm thinking MutableByteSequence could be removed and the functionality collapsed into ArrayByteSequence. That might be a different PR though. I can create an issue for it.

Reply:
Created #4738

Reply:
While it's not technically immutable, I thought the intent was that it was pretty much expected to be used in immutable ways (assuming the backing array didn't change). Isn't that why it's used instead of Text, after all? Otherwise, why not just use Text?

Reply:
There are two big differences between Text and ArrayByteSequence.

 * Text always copies when an array is set on it.  So when doing something like `Text row = key.getRow()` it will allocate a byte array, copy the row into it, and then allocate a text object..  This will not copy the row byte array `ByteSequence row = key.getRowData()`, it will only allocate an ArrayByteSequence object.
 * ArrayByteSequence can refer to a subsequence of a byte array that starts a  position other than zero.  Text always assume 0 start.

The interesting thing about the changes in this PR is that filtering on visibility could do zero object allocations for the common case.  This would not be possible with Text, it will always allocate a byte array and copy.